### ISSUE TITLE
Memory leak in TlsConnectorBuilder::add_root_certificate

### ISSUE THREAD

**Issue by dbrgn** [sfackler/rust-native-tls#74]
I think I found a memory leak. Code to reproduce:

```rust
extern crate native_tls;

use std::fs::File;
use std::io::Read;
use std::path::Path;
use native_tls::{Certificate, TlsConnector};

fn main() {
    println!("Hello, world!");

    let mut cert_bytes = Vec::new();
    File::open(&Path::new("cert.der"))
        .expect("Could not open cert.der")
        .read_to_end(&mut cert_bytes)
        .expect("Could not read cert.der");

    let cert = Certificate::from_der(&cert_bytes).unwrap();

    let mut builder = TlsConnector::builder().unwrap();
    builder.add_root_certificate(cert).unwrap();
    let connector = builder.build();

    // All resources should be dropped here
}
```

Place a DER formatted certificate called `cert.der` in your current directory. Build with `--release`, then run valgrind on it:

```
memtest$ cargo build --release
    Finished release [optimized] target(s) in 0.0 secs
memtest$ valgrind --leak-check=full target/release/memtest 
==10181== Memcheck, a memory error detector
==10181== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==10181== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==10181== Command: target/release/memtest
==10181== 
Hello, world!
==10181== 
==10181== HEAP SUMMARY:
==10181==     in use at exit: 2,344 bytes in 48 blocks
==10181==   total heap usage: 27,300 allocs, 27,252 frees, 2,134,572 bytes allocated
==10181== 
==10181== 2,344 (328 direct, 2,016 indirect) bytes in 1 blocks are definitely lost in loss record 32 of 32
==10181==    at 0x4C2CEDF: malloc (vg_replace_malloc.c:299)
==10181==    by 0x51FD639: CRYPTO_zalloc (in /usr/lib/libcrypto.so.1.1)
==10181==    by 0x5131D07: ??? (in /usr/lib/libcrypto.so.1.1)
==10181==    by 0x512F772: ??? (in /usr/lib/libcrypto.so.1.1)
==10181==    by 0x51302BA: ASN1_item_ex_d2i (in /usr/lib/libcrypto.so.1.1)
==10181==    by 0x513033B: ASN1_item_d2i (in /usr/lib/libcrypto.so.1.1)
==10181==    by 0x110E78: openssl::x509::X509::from_der (in /tmp/memtest/target/release/memtest)
==10181==    by 0x110B31: native_tls::Certificate::from_der (in /tmp/memtest/target/release/memtest)
==10181==    by 0x110436: memtest::main (in /tmp/memtest/target/release/memtest)
==10181==    by 0x110772: std::rt::lang_start::{{closure}} (in /tmp/memtest/target/release/memtest)
==10181==    by 0x113227: {{closure}} (rt.rs:59)
==10181==    by 0x113227: _ZN3std9panicking3try7do_call17hc2d31ff468fce3beE.llvm.A6DD7AAE (panicking.rs:479)
==10181==    by 0x12895E: __rust_maybe_catch_panic (lib.rs:102)
==10181== 
==10181== LEAK SUMMARY:
==10181==    definitely lost: 328 bytes in 1 blocks
==10181==    indirectly lost: 2,016 bytes in 47 blocks
==10181==      possibly lost: 0 bytes in 0 blocks
==10181==    still reachable: 0 bytes in 0 blocks
==10181==         suppressed: 0 bytes in 0 blocks
==10181== 
==10181== For counts of detected and suppressed errors, rerun with: -v
==10181== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

The leak does not occur if I comment out the `add_root_certificate` line.

Not sure if this is in the `native-tls` or the `openssl` crate.

**Comment by sfackler** [comment]
The problem would be in the openssl crate - I'll take a look now!

**Comment by sfackler** [comment]
https://github.com/sfackler/rust-openssl/pull/836

**Comment by sfackler** [comment]
Fix is in openssl 0.9.24.

**Comment by dbrgn** [comment]
Thanks!
